lib/portlayer/exec/commit.go

```diff
@@ -0,0 +1,182 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exec
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/lib/portlayer/event/events"
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/vsphere/session"
+	"github.com/vmware/vic/pkg/vsphere/tasks"
+	"github.com/vmware/vic/pkg/vsphere/vm"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// Commit executes the requires steps on the handle
+func Commit(ctx context.Context, sess *session.Session, h *Handle, waitTime *int32) error {
+	defer trace.End(trace.Begin(h.ExecConfig.ID))
+
+	c := Containers.Container(h.ExecConfig.ID)
+	creation := h.vm == nil
+	if creation {
+		if h.Spec == nil {
+			return fmt.Errorf("a spec must be provided for create operations")
+		}
+
+		if sess == nil {
+			// session must not be nil
+			return fmt.Errorf("no session provided for create operations")
+		}
+
+		// the only permissible operation is to create a VM
+		if h.Spec == nil {
+			return fmt.Errorf("only create operations can be committed without an existing VM")
+		}
+
+		if c != nil {
+			return fmt.Errorf("a container already exists in the cache with this ID")
+		}
+
+		var res *types.TaskInfo
+		var err error
+		if sess.IsVC() && Config.VirtualApp.ResourcePool != nil {
+			// Create the vm
+			res, err = tasks.WaitForResult(ctx, func(ctx context.Context) (tasks.Task, error) {
+				return Config.VirtualApp.CreateChildVM_Task(ctx, *h.Spec.Spec(), nil)
+			})
+		} else {
+			// Find the Virtual Machine folder that we use
+			var folders *object.DatacenterFolders
+			folders, err = sess.Datacenter.Folders(ctx)
+			if err != nil {
+				log.Errorf("Could not get folders")
+				return err
+			}
+			parent := folders.VmFolder
+
+			// Create the vm
+			res, err = tasks.WaitForResult(ctx, func(ctx context.Context) (tasks.Task, error) {
+				return parent.CreateVM(ctx, *h.Spec.Spec(), Config.ResourcePool, nil)
+			})
+		}
+
+		if err != nil {
+			log.Errorf("Something failed. Spec was %+v", *h.Spec.Spec())
+			return err
+		}
+
+		h.vm = vm.NewVirtualMachine(ctx, sess, res.Result.(types.ManagedObjectReference))
+		c = newContainer(&h.containerBase)
+		Containers.Put(c)
+		// inform of creation irrespective of remaining operations
+		publishContainerEvent(c.ExecConfig.ID, time.Now().UTC(), events.ContainerCreated)
+
+		// clear the spec as we've acted on it - this prevents a reconfigure from occurring in follow-on
+		// processing
+		h.Spec = nil
+	}
+
+	// if we're stopping the VM, do so before the reconfigure to preserve the extraconfig
+	poweroff := false
+	if h.TargetState() == StateStopped {
+		if h.Runtime == nil {
+			log.Warnf("Commit called with incomplete runtime state for %s", h.ExecConfig.ID)
+		}
+
+		if h.Runtime != nil && h.Runtime.PowerState == types.VirtualMachinePowerStatePoweredOff {
```

Comment:
The logic is fine, once we have ChangeVersion enforcing things - we have to assume that the changes represented by the handle are only valid if the state of the container hasn't changed in the interim. Longer term we could add a diff'ing mechanism to the changes to become more granular about it.

For now I've made it so that there's always an explicit refresh of state before the reconfigure - this removes the protection of the ChangeVersion but avoids potential issues with reconfiguring a liveVM